instrumentation/jdbi3/src/main/java/brave/jdbi3/BraveStatementContextListener.java

```diff
@@ -0,0 +1,142 @@
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.time.Instant;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.jdbi.v3.core.statement.SqlStatements;
+import org.jdbi.v3.core.statement.StatementContext;
+import org.jdbi.v3.core.statement.StatementContextListener;
+
+import brave.Span;
+import brave.Tracing;
+
+public class BraveStatementContextListener implements StatementContextListener {
+    private final Tracing tracing;
+    private static final Pattern serviceNamePattern = Pattern.compile("(^|&)zipkinServiceName=(?<serviceName>[^&]*)");
+    private Map<String, String> serviceNameCache = new ConcurrentHashMap<String, String>() { };
+
+    public BraveStatementContextListener(Tracing tracing) { this.tracing = tracing; }
+
+    @Override
+    public void contextCreated(final StatementContext ctx) {
+        String spanName = ctx.describeJdbiStatementType();
+        Span span = tracing.tracer().nextSpan().name(spanName);
+
+        Instant start = ctx.getExecutionMoment();
+        if (start != null) {
+            span.start(start.toEpochMilli() * 1000);
+        } else {
+            span.start();
+        }
+
+        ctx.setTraceId(span.context().traceIdString());
+
+        ctx.addCleanable(() -> {
+            final SqlStatements stmtConfig = ctx.getConfig(SqlStatements.class);
+
+            String remoteServiceName = getRemoteServiceName(ctx);
+            if (remoteServiceName != null && !remoteServiceName.isEmpty()) {
+                span.remoteServiceName(remoteServiceName);
+            }
+
+            span.kind(Span.Kind.CLIENT);
+
+            final String renderedSql = ctx.getRenderedSql();
+            if (renderedSql != null) {
+                String truncated = renderedSql.substring(
+                    0,
+                    Math.min(renderedSql.length(), stmtConfig.getJfrSqlMaxLength())
+                );
+                span.tag("sql.query", truncated);
+            }
+
+            span.tag("sql.rows", Long.toString(ctx.getMappedRows()));
+
+            if (ctx.getCompletionMoment() == null) {
+                span.tag("error", "");
+            }
+
+            if (ctx.getCompletionMoment() != null) {
+                span.finish(ctx.getCompletionMoment().toEpochMilli() * 1000);
+            } else if (ctx.getExceptionMoment() != null) {
+                span.finish(ctx.getExceptionMoment().toEpochMilli() * 1000);
+            } else {
+                span.finish();
+            }
+        });
+    }
+
+    String getRemoteServiceName(StatementContext ctx) {
+        Connection connection = ctx.getConnection();
+        if (connection == null) {
+            return null;
+        }
+        URI uri;
+        try {
+            String url = connection.getMetaData().getURL();
+            if (serviceNameCache.containsKey(url)) {
```

Comment:
Since we use `ConcurrentHashMap` now:
```suggestion
        return serviceNameCache.computeIfAbsent(url, key -> {
                if (url.startsWith("jdbc:")) {
                    // strip "jdbc:" prefix
                    url = url.substring(5);
                }
                final URI uri = new URI(url); // strip "jdbc:"
                return  extractRemoteServiceName(uri);
            });
```